Add documentation on request cancellation (#6403)
triton-inference-server/server: docs/user_guide/request_cancellation.md

and `TRITONSERVER_InferenceRequestIsCancelled` to cancel and query the cancellation
status of an inflight request. Read more about the APIs in [tritonserver.h](https://github.com/triton-inference-server/core/blob/main/include/triton/core/tritonserver.h).

In addition, [gRPC endpoint](../customization_guide/inference_protocols.md#httprest-and-grpc-protocols) can
now detect cancelation from the client and attempt to terminate request.
At present, only gRPC python client supports issuing request cancellation
to the server endpoint. See [request-cancellation](https://github.com/triton-inference-server/client#request-cancellation)
for more details on how to issue requests from the client-side.
See gRPC guide on RPC [cancellation](https://grpc.io/docs/guides/cancellation/) for
finer details. 

Upon receiving request cancellation, triton does its best to cancel request
at various points. However, once a request has been given to the backend
for execution, it is upto the individual backends to detect and handle
request termination.

> Currently, following backend(s) support(s) early termination:

Reviewer: Are we not mentioning Python backend here?

Author: Fixed for consistency and added special case: https://github.com/triton-inference-server/server/pull/6403/commits/99cb38e97ee0efa2e1c4d8ec069b585d305c253d

and later determine the results are no longer needed. Continuing to process
requests whose results are no longer required can significantly impact server
resources.

[In-Process Triton Server C API](../customization_guide/inference_protocols.md#in-process-triton-server-api) has been enhanced with `TRITONSERVER_InferenceRequestCancel`
and `TRITONSERVER_InferenceRequestIsCancelled` to cancel and query the cancellation
status of an inflight request. Read more about the APIs in [tritonserver.h](https://github.com/triton-inference-server/core/blob/main/include/triton/core/tritonserver.h).

In addition, [gRPC endpoint](../customization_guide/inference_protocols.md#httprest-and-grpc-protocols) can
now detect cancellation from the client and attempt to terminate request.
At present, only gRPC python client supports issuing request cancellation
to the server endpoint. See [request-cancellation](https://github.com/triton-inference-server/client#request-cancellation)
for more details on how to issue requests from the client-side.
See gRPC guide on RPC [cancellation](https://grpc.io/docs/guides/cancellation/) for
finer details.

> Upon receiving request cancellation, triton does its best to cancel request

Reviewer 2: Can we expand on some core/scheduler based cancellation semantics? For example, should we mention the behavior around sequences when cancelling a request in a sequence?

Reviewer 2: Maybe we can make 3 subsections to this section for Frontend, Core, Backend

at various points. However, once a request has been given to the backend
for execution, it is upto the individual backends to detect and handle
request termination.

> Currently, following backend(s) support(s) early termination:

Reviewer 2: ```suggestion
Currently, the following backends support early termination:
```

    - [vLLM backend](https://github.com/triton-inference-server/vllm_backend)
    - [python backend](https://github.com/triton-inference-server/python_backend)

Python backend is a special case where we expose the APIs to detect cancellation

> status of the request but it is upto the `model.py` developer to detect whether

Reviewer 3: ```suggestion
status of the request but it is up to the `model.py` developer to detect whether
```

the request is cancelled and terminate further execution.

**For the backend developer**: The backend APIs have also been enhanced to let the
backend detect whether the request received from Triton core has been cancelled.
See `TRITONBACKEND_RequestIsCancelled` and `TRITONBACKEND_ResponseFactoryIsCancelled`
in [tritonbackend.h](https://github.com/triton-inference-server/core/blob/main/include/triton/core/tritonbackend.h)
for more details. The backend upon detecting request cancellation can stop processing
it any further.

> The python models running behind python backend can also query the cancellation status

Reviewer 4: ```suggestion
The Python models running behind Python backend can also query the cancellation status
```

requests whose results are no longer required can significantly impact server
resources.

## Issuing Request Cancellation

### Triton C API

[In-Process Triton Server C API](../customization_guide/inference_protocols.md#in-process-triton-server-api) has been enhanced with `TRITONSERVER_InferenceRequestCancel`
and `TRITONSERVER_InferenceRequestIsCancelled` to cancel and query the cancellation

> status of an inflight request. Read more about the APIs in [tritonserver.h](https://github.com/triton-inference-server/core/blob/main/include/triton/core/tritonserver.h).

Reviewer: ```suggestion
[In-Process Triton Server C API](../customization_guide/inference_protocols.md#in-process-triton-server-api)
has been enhanced with `TRITONSERVER_InferenceRequestCancel` and
`TRITONSERVER_InferenceRequestIsCancelled` to issue cancellation and query
cancellation issuance of an inflight request. Read more about the APIs in
[tritonserver.h](https://github.com/triton-inference-server/core/blob/main/include/triton/core/tritonserver.h).
```

See gRPC guide on RPC [cancellation](https://grpc.io/docs/guides/cancellation/) for
finer details.

## Handling in Triton Core

Triton core checks for requests that have been cancelled at some critical points
when using [dynamic](./model_configuration.md#dynamic-batcher) or
[sequence batching](./model_configuration.md#sequence-batcher). We also test for
the cancelled requests after every [ensemble](./model_configuration.md#ensemble-scheduler)
step and terminate further processing the requests.

On detecting a cancelled request, Triton core responds with CANCELLED status. If a request
is cancelled when using [sequence_batching](./model_configuration.md#sequence-batcher),
then all the pending requests in the same sequence will also be cancelled. The sequence
is represented by the requests that has identical sequence id.

**Note**: Currently, Triton core does not detect cancellation status of a request once
it is forwarded to [rate limiter](./rate_limiter.md). Improving the request cancellation
detection and handling within Triton core is work in progress.

## Handling in Backend

> Upon receiving request cancellation, triton does its best to terminate request

Reviewer 2: Double check how the bullet points below are rending, might be wrong

Author: Good catch! Fixed,
